File: BackEnd/PHP/laravelshop/app/Http/Controllers/Api/V1/Ranking/RankingController.php
<?php

namespace App\Http\Controllers\Api\V1\Ranking;

use App\Http\Controllers\Api\V1\ApiController;
use App\Modules\Ranking\Services\RankingService;

class RankingController extends ApiController
{
    public function __construct(
        private readonly RankingService $rankingService,
    ) {
    }

    public function products()
    {
        return $this->success($this->rankingService->popularProducts(
            request()->query('categoryId') ? (int) request()->query('categoryId') : null,
            (int) request()->query('limit', 20),
        ));
    }

    public function keywords()
    {
        return $this->success($this->rankingService->popularKeywords((int) request()->query('limit', 20)));
    }

    public function recalculate()
    {
        return $this->success($this->rankingService->recalculate(request()->user()));
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/ranking/RankingController.java
package com.pbshop.java.spring.maven.jpa.postgresql.ranking;

import java.util.List;
import java.util.Map;

import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.common.api.ApiResponse;

@RestController
@RequestMapping("${pbshop.api.base-path:/api/v1}/rankings")
public class RankingController {

    private final RankingService rankingService;

    public RankingController(RankingService rankingService) {
        this.rankingService = rankingService;
    }

    @GetMapping("/products/popular")
    public ApiResponse<List<Map<String, Object>>> getPopularProducts(
            @RequestParam(required = false) Long categoryId,
            @RequestParam(required = false) String period,
            @RequestParam(required = false) Integer limit
    ) {
        return ApiResponse.success(rankingService.getPopularProducts(categoryId, limit));
    }

    @GetMapping("/keywords/popular")
    public ApiResponse<List<Map<String, Object>>> getPopularKeywords(
            @RequestParam(required = false) Integer limit
    ) {
        return ApiResponse.success(rankingService.getPopularKeywords(limit));
    }

    @PostMapping("/admin/recalculate")
    public ApiResponse<Map<String, Object>> recalculate(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal
    ) {
        return ApiResponse.success(rankingService.recalculate(principal));
    }
}
